elastic/beats
# Adding functionality to death letter index instead of dropping events.

## What does this PR do?

Allows the user to specify the policy (/action) of how to deal with non-indexable events. We default to `drop` dropping the event which is current behavior. Alternatively one can specify `death_letter_index` to index into the specified index name

```
output.elasticsearch:
    non_indexable_policy.dead_letter_index:
        index: "dead-letter-index"
```

```
output.elasticsearch:
    non_indexable_policy.drop: ~
```

## Why is it important?

Currently when there's a mapping conflict events are dropped, this is undesirable for a lot of customers. While filebeat technically doesn't violate it's guarantee that events are delivered at least once, it does cause the event to be lost. Which is both undesirable and unexpected behavior for most customers.

## Checklist


## How to test this PR locally

See integration test, 
Configure filebeat with the ES ouput and specifying the action and index
Then creating a mapping conflict:
```
{ "bar": 0 }
{ "bar": "bar0" }
```

## Related issues

## Use cases

When a mapping conflict occurs the message is kept and indexed into an alternative index. This can then later be re-ingested after either the data or the index (the alias points to) is changed. 

## Other remarks

- The index currently must reside on the same ES cluster. Changing this, if desired, would be a separate PR. 
- The format we use to store the events on the death letter index is message fields (containing the original message with escaped json), `error.type` (http code) and `error.message` (description)
